fairscale/optim/oss.py

```diff
@@ -310,12 +297,116 @@ def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
+        sharded properties are not exposed. It contains two entries:
+
+        * state - a dict holding current optimization state. Its content
+            differs between optimizer classes.
+
+        * param_groups - a dict containing all parameter groups
+
+        .. warning:
+            If the state has not been consolidated, this returns a shard's worth, not the global state.
+
+        .. warning:
+            Returning the global state is limited to the replica which was responsible for the consolidation.
+            The state may also not be up to date, depending on when `consolidate_state_dict` was last called.
+        """
+
+        if len(self._all_states) == 0:
+            raise RuntimeError(
+                "Optimizer state has not been consolidated on this rank. \
+                Please call `consolidate_state_dict()` on all ranks beforehand if you meant to save the global state"
+            )
+
+        # Unify the shard states and the state that pytorch would expect, given the model.
+        # Indexation needs several redirections, since each shard only knows a limited scope of the model
+        # - get the pytorch compliant parameter indexing
+        state_dict = super().state_dict()
+
+        # - get an id map which links the parameter id to the index in the reference state
+        global_id_map = {id(p): i for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))}
+
+        # - go through the per-shard states, which are all indexed locally
+        for rank, s in enumerate(self._all_states):
+            # -- match the local indexing and the global partition, update the corresponding saved state globally
+            for local_pg, global_pg in zip(s["param_groups"], self.partition_parameters()[rank]):
+                local_index_to_param_id = {
+                    i_param: id(global_pg["params"][i]) for i, i_param in enumerate(local_pg["params"])
+                }
+
+                for local_param_index in local_pg["params"]:
+                    # Update the state, if any
+                    if local_param_index in s["state"].keys():
+                        global_id = global_id_map[local_index_to_param_id[local_param_index]]
+                        state_dict["state"][global_id] = s["state"][local_param_index]
+
+        return state_dict
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """Restore the global parameter groups as well as the shard.
+
+        Arguments:
+            state_dict (dict): optimizer state. Should be an object returned
+                from a call to :meth:`state_dict`
+        """
+
+        # NOTE: PyTorch 1.5 does not index linearly but with the id(params) at saving time
+        # we work around that here by using the fact that the params are ordered as in the param_groups
+
+        # Param index to param map
+        global_id_map = {i: p for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))}
```

Comment:
what if the user adds param_groups ? then it needs to be updated. It's a one liner, but it can be moved somewhere else , of course

Reply:
I think adding param_groups after __init__ is undefined behavior. I don't think this is code is robust to param_groups being added for other reasons.

In addition to the code duplication, iterating over all params seems like it would be expensive for a large model so it might be worth the savings to do it just once.

Reply:
There's a optim.add_param_group() method, I'm a bit surprised that this would lead to undefined behavior, I'm not sure that we're talking about the same thing. Else sure I can cache this somewhere else, hard to guess what people prefer (it will add an indirection), but sure, I will do that.

Reply:
https://pytorch.org/docs/stable/optim.html#torch.optim.Optimizer.add_param_group

Reply:
updated